File: src/algorithms/cusum.ts
export interface CUSUMOptions {
	target: number;
	slack: number;
	threshold: number;
}

export class CUSUM {
	private target: number;
	private slack: number;
	private threshold: number;
	private cPos: number = 0;
	private cNeg: number = 0;
	public alarm: boolean = false;
	public alarmType: "high" | "low" | null = null;
	public alarmValue: number | null = null;

	constructor(options: CUSUMOptions) {
		this.target = options.target;
		this.slack = options.slack;
		this.threshold = options.threshold;
	}

	update(sample: number): boolean {
		this.alarm = false;
		this.alarmType = null;
		this.alarmValue = null;
		
		const deviation = sample - this.target;
		
		// Upper CUSUM: Accumulate positive deviations above slack
		this.cPos = Math.max(0, this.cPos + deviation - this.slack);
		
		// Lower CUSUM: Accumulate negative deviations (deviation is negative, so -deviation is positive)
		this.cNeg = Math.max(0, this.cNeg - deviation - this.slack);

		if (this.cPos > this.threshold) {
			this.alarm = true;
			this.alarmType = "high";
			this.alarmValue = this.cPos;
			// Optional: Reset partially or fully after alarm. 
			// Standard CUSUM resets to 0.
			this.cPos = 0; 
			return true;
		}

		if (this.cNeg > this.threshold) {
			this.alarm = true;
			this.alarmType = "low";
			this.alarmValue = this.cNeg;
			this.cNeg = 0;
			return true;
		}

		return false;
	}

	setTarget(target: number): void {
		this.target = target;
	}

	setParameters(target: number, slack: number, threshold: number): void {
		this.target = target;
		this.slack = slack;
		this.threshold = threshold;
	}

	getValues(): { upper: number; lower: number } {
		return { upper: this.cPos, lower: this.cNeg };
	}

	reset(): void {
		this.cPos = 0;
		this.cNeg = 0;
		this.alarm = false;
		this.alarmType = null;
		this.alarmValue = null;
	}
}

File: via-core/crates/via-core/src/algo/cusum.rs
pub struct CUSUM {
    target: f64,
    slack: f64,
    threshold: f64,
    c_pos: f64,
    c_neg: f64,
    pub alarm: bool,
    pub alarm_type: i8, // 0=none, 1=high, -1=low
}

impl CUSUM {
    pub fn new(target: f64, slack: f64, threshold: f64) -> Self {
        Self {
            target,
            slack,
            threshold,
            c_pos: 0.0,
            c_neg: 0.0,
            alarm: false,
            alarm_type: 0,
        }
    }

    pub fn update(&mut self, sample: f64) -> bool {
        self.alarm = false;
        self.alarm_type = 0;

        let deviation = sample - self.target;

        self.c_pos = (self.c_pos + deviation - self.slack).max(0.0);
        self.c_neg = (self.c_neg - deviation - self.slack).max(0.0);

        if self.c_pos > self.threshold {
            self.alarm = true;
            self.alarm_type = 1;
            self.c_pos = 0.0; // Reset after alarm
            return true;
        }

        if self.c_neg > self.threshold {
            self.alarm = true;
            self.alarm_type = -1;
            self.c_neg = 0.0;
            return true;
        }

        false
    }
}
